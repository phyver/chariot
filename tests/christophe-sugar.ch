(**********************************************************************
 *** tests with nested mu nu / nu mu / nu nu
 *********************************************************************)

:set verbose 1
:set allow_structs true
:set keep_useless_clauses true

--------------
-- basic types
data nat where
    | Zero : nat
    | Succ : nat -> nat

data plus('a,'b) where
    | Cons0 : 'a -> plus('a,'b)
    | Cons1 : 'b -> plus('a,'b)

codata stream('a) where
    | Head : stream('a) -> 'a
    | Tail : stream('a) -> stream('a)

-- X  |-->  nu Y . (X + Y)
codata nuY('x) where
    | Nu : nuY('x) -> plus('x,nuY('x))

-- X  |-->  mu Y . (X + Y)
data muY('x) where
    | Mu : plus('x,muY('x)) -> muY('x)


------------------------
-- mu X . mu Y . (X + Y)
-- this type is empty
data list1 where
    | Cons : muY(list1) -> list1

------------------------
-- nu X . mu Y . (X + Y)
-- infinite streams, with only finitely many consecutive 1s
codata stream2 where
    | D2 : stream2 -> muY(stream2)

------------------------
-- mu X . nu Y . (X + Y)
-- infinite streams, infinite number of 1s at the end (constructor Cons1)
data stream3 where
    | C3 : nuY(stream3) -> stream3

------------------------
-- nu x . nu y . (x + y)
-- infinite streams of 0 / 1 (constructors Cons0 / Cons1)
codata stream4 where
    | D4 : stream4 -> nuY(stream4)


-------------------------------------------------------
-- converting the different kinds of streams to "usual"
-- streams of natural numbers
val convert2 : stream2 -> stream(nat)
  | convert2 (D2 # Mu $ Cons0 s) = { Head = 0 ; Tail = convert2 s }
  | convert2 (D2 # Mu $ Cons1 s) = { Head = 1 ; Tail = convert2 $ D2 # s }

val convert3 : stream3 -> stream(nat)
  | convert3 (C3 $ Nu # Cons0 s) = { Head = 0 ; Tail = convert3 s }
  | convert3 (C3 $ Nu # Cons1 s) = { Head = 1 ; Tail = convert3 $ C3 s }


val convert4 : stream4 -> stream(nat)
  | convert4 (D4 # Nu # Cons0 s) = { Head = 0 ; Tail = convert4 s }
  | convert4 (D4 # Nu # Cons1 s) = { Head = 1 ; Tail = convert4 $ D4 # s }



-------------------------------------------------------
-- converting between the different kinds of streams

-- from stream2 to stream4
val stream2_to_4 : stream2 -> stream4
  | stream2_to_4 (D2 # Mu $ Cons0 s) = D4 # Nu # Cons0 $ stream2_to_4 s
  | stream2_to_4 (D2 # Mu $ Cons1 s) = D4 # Nu # Cons1 (stream2_to_4 $ D2 # s).D4

-- from stream3 to stream4
val stream3_to_4 : stream3 -> stream4
  | stream3_to_4 (C3 $ Nu # Cons0 s) = D4 # Nu # Cons0 $ stream3_to_4 s
  | stream3_to_4 (C3 $ Nu # Cons1 s) = D4 # Nu # Cons1 $ s34_map s
and
    s34_map : nuY(stream3) -> nuY(stream4)
  | s34_map (Nu # Cons0 s) = Nu # Cons0 $ stream3_to_4 s
  | s34_map (Nu # Cons1 s) = Nu # Cons1 $ s34_map s

-- TODO: the SCT doesn't see that this version terminates
-- it seems to be correct because the second rule cannot be applied infinitely often...
val stream3_to_4_bis : stream3 -> stream4
  | stream3_to_4_bis (C3 $ Nu # Cons0 s) = D4 # Nu # Cons0 $ stream3_to_4_bis s
  | stream3_to_4_bis (C3 $ Nu # Cons1 s) = D4 # Nu # Cons1 $ (stream3_to_4_bis $ C3 s).D4

-- from stream3 to stream2, replacing 1s by 0s
val stream3_to_2_op : stream3 -> stream2
  | stream3_to_2_op (C3 $ Nu # Cons0 s) = D2 # Mu $ Cons1 $ (stream3_to_2_op s).D2
  | stream3_to_2_op (C3 $ Nu # Cons1 s) = D2 # Mu $ Cons0 $ stream3_to_2_op $ C3 s


-------------------------------------------------------
-- constructing streams

-- only 0s
val zeros2 : stream2
  | zeros2 = D2 # Mu $ Cons0 zeros2


-- construct 1110 1110 1110 1110 ... where each block contains n 1s
val
    ones_n2_aux : nat -> nat -> stream2
  | ones_n2_aux 0 n = D2 # Mu $ Cons0 $ ones_n2_aux n n
  | ones_n2_aux (k+1) n = D2 # Mu $ Cons1 (ones_n2_aux k n).D2

val ones_n2 : nat -> stream2
  | ones_n2 n = ones_n2_aux n n


-- construct 01...1 01...1 01...1 ... where the i-th block of 1s contains (f i) 1s
val
    ones_f2_aux : (nat->nat) -> nat -> nat -> stream2
  | ones_f2_aux f 0 i = D2 # Mu $ Cons0 $ ones_f2_aux f (f i) (i+1)
  | ones_f2_aux f (k+1) i = D2 # Mu $Cons1 (ones_f2_aux f k i).D2

val ones_f2 : (nat -> nat) -> stream2
  | ones_f2 f = ones_f2_aux f 0 0

-- only 1s
val ones3_aux : nuY(stream3)
  | ones3_aux = Nu # Cons1 ones3_aux

val ones3 : stream3
  | ones3 = C3 ones3_aux


-- some 0s, then only 1s
val n_zeros_ones3 : nat -> stream3
  | n_zeros_ones3 0 = ones3
  | n_zeros_ones3 (n+1) = C3 (n_zeros_ones3_aux n)
and n_zeros_ones3_aux : nat -> nuY(stream3)
  | n_zeros_ones3_aux n = Nu # Cons0 $ n_zeros_ones3 n


-- start like a given stream2 for some time, then only ones
val get_n_ones3 : nat -> stream2 -> stream3
  | get_n_ones3 n s = C3 (get_n_ones3_aux n s.D2)
and
    get_n_ones3_aux : nat -> muY(stream2) -> nuY(stream3)
  | get_n_ones3_aux 0 s = ones3_aux
  | (get_n_ones3_aux (n+1) (Mu (Cons0 s))) = Nu # Cons0 $ get_n_ones3 n s
  | (get_n_ones3_aux (n+1) (Mu (Cons1 s))) = Nu # Cons1 $ get_n_ones3_aux n s

-- get the length of blocks of 1s
val
    map_length1 : stream2 -> stream(nat)
  | map_length1 s = map_length1_aux s.D2 0
and
    map_length1_aux : muY(stream2) -> nat -> stream(nat)
  | (map_length1_aux (Mu (Cons0 s)) n) = { Head = n ; Tail = map_length1 s }
  | (map_length1_aux (Mu (Cons1 s)) n) = map_length1_aux s (n+1)


-- tests

-- only zeros
val test1 = convert2 zeros2
:unfold test1, 3

-- blocs of 3 1s,
val test2 = convert2 (ones_n2 3)
:unfold test2, 9

-- blocs of 3 1s, other version
val k3 n = 3
val test3 = convert2 (ones_f2 k3)
:unfold test3, 9

-- blocs of 1s of increasing length
val id n = n
val test4 = convert2 (ones_f2 id)
:unfold test4, 16

-- only 1s
val test5 = convert3 (ones3)
:unfold test5, 6

-- 5 0s,then only 1s
val test6 = convert3 (n_zeros_ones3 5)
:unfold test6, 9

-- increasing blocks of 1s, then only 1s
val test7 = convert3 (get_n_ones3 11 (ones_f2 id))
:unfold test7, 20

-- length of the blocs of ones
val test8 = map_length1 (ones_n2 3)
:unfold test8, 10

val test9 =map_length1 (ones_f2 k3)
:unfold test9, 10

val test10 = map_length1 (ones_f2 id)
:unfold test10, 10

-- blocs of 1s of increasing length, after conversion to stream4
val test12 = convert4 (stream2_to_4 (ones_f2 id))
:unfold test12, 16

-- increasing blocks of 1s, then only 1s, after conversion to stream4
val test13 = convert4 (stream3_to_4 (get_n_ones3 11 (ones_f2 id)))
:unfold test13, 20

-- five 1s, then only 0s
val test14 = convert2 (stream3_to_2_op (n_zeros_ones3 5))
:unfold test14, 12


(*
-- other, malfunctioning definitions
val
    cons3_aux : nat -> nuY(stream3) -> nuY(stream3)
  | (cons3_aux 0 s).Nu = Cons0 (C3 s)
  | (cons3_aux 1 s).Nu = Cons1 s
  | (cons3_aux n s).Nu = !!!
 val cons3 : nat -> stream3 -> stream3
   | cons3 k (C3 s) = C3 (cons3_aux k s)

-- doesn't terminate
val ones3_bad : stream3
  | ones3_bad = cons3 1 ones3_bad


val map_nuY : ('a -> 'b) -> nuY('a) -> nuY('b)
  | (map_nuY f s).Nu = map_nuY_aux f s.Nu
and
    map_nuY_aux : ('a -> 'b) -> plus('a,nuY('a)) -> plus('b,nuY('b))
  | map_nuY_aux f (Cons0 a) = Cons0 (f a)
  | map_nuY_aux f (Cons1 s) = Cons1 (map_nuY f s)

-- the next version doesn't pass the SCT because of recursive call via map_nuY
val stream3_to_4_bad: stream3 -> stream4
  | stream3_to_4_bad (C3 s) = stream3_to_4_bad_aux s.Nu
and stream3_to_4_bad_aux : plus(stream3,nuY(stream3)) -> stream4
  | (stream3_to_4_bad_aux (Cons0 s)).D4.Nu = Cons0 (stream3_to_4_bad s)
  | (stream3_to_4_bad_aux (Cons1 s)).D4.Nu = Cons1 (map_nuY stream3_to_4_bad s)
*)

