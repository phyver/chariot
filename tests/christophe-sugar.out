>> inductive type nat was succesfully defined
>> inductive type plus was succesfully defined
>> coinductive type stream was succesfully defined
>> coinductive type nuY was succesfully defined
>> inductive type muY was succesfully defined
>> inductive type list1 was succesfully defined
>> coinductive type stream2 was succesfully defined
>> inductive type stream3 was succesfully defined
>> coinductive type stream4 was succesfully defined
>> Typing for _aux₄, _aux₃, _aux₂, _aux₁, convert2 successful
>> the definition for convert2 is complete
>> the definition for _aux₁ is complete
>> the definition for _aux₂ is complete
>> the definition for _aux₃ is complete
>> the definition for _aux₄ is complete
>> the definitions _aux₄, _aux₃, _aux₂, _aux₁, convert2 are provably correct

>> Typing for _aux₇, _aux₆, _aux₅, convert3 successful
>> the definition for convert3 is complete
>> the definition for _aux₅ is complete
>> the definition for _aux₆ is complete
>> the definition for _aux₇ is complete
>> the definitions _aux₇, _aux₆, _aux₅, convert3 are provably correct

>> Typing for _aux₉, _aux₈, _aux₁₂, _aux₁₁, _aux₁₀, convert4 successful
>> the definition for convert4 is complete
>> the definition for _aux₁₀ is complete
>> the definition for _aux₁₁ is complete
>> the definition for _aux₁₂ is complete
>> the definition for _aux₈ is complete
>> the definition for _aux₉ is complete
>> the definitions _aux₉, _aux₈, _aux₁₂, _aux₁₁, _aux₁₀, convert4 are provably correct

>> Typing for _aux₁₈, _aux₁₇, _aux₁₆, _aux₁₅, _aux₁₄, _aux₁₃, stream2_to_4 successful
>> the definition for stream2_to_4 is complete
>> the definition for _aux₁₃ is complete
>> the definition for _aux₁₄ is complete
>> the definition for _aux₁₅ is complete
>> the definition for _aux₁₆ is complete
>> the definition for _aux₁₇ is complete
>> the definition for _aux₁₈ is complete
>> the definitions _aux₁₈, _aux₁₇, _aux₁₆, _aux₁₅, _aux₁₄, _aux₁₃, stream2_to_4 are provably correct

>> Typing for _aux₂₆, _aux₂₅, _aux₂₄, _aux₂₃, _aux₂₂, _aux₂₁, _aux₂₀, _aux₁₉, s34_map, stream3_to_4 successful
>> the definition for stream3_to_4 is complete
>> the definition for s34_map is complete
>> the definition for _aux₁₉ is complete
>> the definition for _aux₂₀ is complete
>> the definition for _aux₂₁ is complete
>> the definition for _aux₂₂ is complete
>> the definition for _aux₂₃ is complete
>> the definition for _aux₂₄ is complete
>> the definition for _aux₂₅ is complete
>> the definition for _aux₂₆ is complete
>> the definitions _aux₂₆, _aux₂₅, _aux₂₄, _aux₂₃, _aux₂₂, _aux₂₁, _aux₂₀, _aux₁₉, s34_map, stream3_to_4 are provably correct

>> Typing for _aux₃₁, _aux₃₀, _aux₂₉, _aux₂₈, _aux₂₇, stream3_to_4_bis successful
>> the definition for stream3_to_4_bis is complete
>> the definition for _aux₂₇ is complete
>> the definition for _aux₂₈ is complete
>> the definition for _aux₂₉ is complete
>> the definition for _aux₃₀ is complete
>> the definition for _aux₃₁ is complete
!! warning: the definitions _aux₃₁, _aux₃₀, _aux₂₉, _aux₂₈, _aux₂₇, stream3_to_4_bis are NOT provably correct (weight_bound: 2, depth_bound: 2)

>> Typing for _aux₃₄, _aux₃₃, _aux₃₂, stream3_to_2_op successful
>> the definition for stream3_to_2_op is complete
>> the definition for _aux₃₂ is complete
>> the definition for _aux₃₃ is complete
>> the definition for _aux₃₄ is complete
>> the definitions _aux₃₄, _aux₃₃, _aux₃₂, stream3_to_2_op are provably correct

>> Typing for _aux₃₅, zeros2 successful
>> the definition for zeros2 is complete
>> the definition for _aux₃₅ is complete
>> the definitions _aux₃₅, zeros2 are provably correct

>> Typing for _aux₃₇, _aux₃₆, ones_n2_aux successful
>> the definition for ones_n2_aux is complete
>> the definition for _aux₃₆ is complete
>> the definition for _aux₃₇ is complete
>> the definitions _aux₃₇, _aux₃₆, ones_n2_aux are provably correct

>> Typing for ones_n2 successful
>> the definition for ones_n2 is complete
>> the definition ones_n2 is provably correct

>> Typing for _aux₃₉, _aux₃₈, ones_f2_aux successful
>> the definition for ones_f2_aux is complete
>> the definition for _aux₃₈ is complete
>> the definition for _aux₃₉ is complete
>> the definitions _aux₃₉, _aux₃₈, ones_f2_aux are provably correct

>> Typing for ones_f2 successful
>> the definition for ones_f2 is complete
>> the definition ones_f2 is provably correct

>> Typing for _aux₄₀, ones3_aux successful
>> the definition for ones3_aux is complete
>> the definition for _aux₄₀ is complete
>> the definitions _aux₄₀, ones3_aux are provably correct

>> Typing for ones3 successful
>> the definition for ones3 is complete
>> the definition ones3 is provably correct

>> Typing for _aux₄₁, n_zeros_ones3_aux, n_zeros_ones3 successful
>> the definition for n_zeros_ones3 is complete
>> the definition for n_zeros_ones3_aux is complete
>> the definition for _aux₄₁ is complete
>> the definitions _aux₄₁, n_zeros_ones3_aux, n_zeros_ones3 are provably correct

>> Typing for _aux₄₃, _aux₄₂, get_n_ones3_aux, get_n_ones3 successful
>> the definition for get_n_ones3 is complete
>> the definition for get_n_ones3_aux is complete
>> the definition for _aux₄₂ is complete
>> the definition for _aux₄₃ is complete
>> the definitions _aux₄₃, _aux₄₂, get_n_ones3_aux, get_n_ones3 are provably correct

>> Typing for _aux₄₄, map_length1_aux, map_length1 successful
>> the definition for map_length1 is complete
>> the definition for map_length1_aux is complete
>> the definition for _aux₄₄ is complete
>> the definitions _aux₄₄, map_length1_aux, map_length1 are provably correct

>> Typing for test1 successful
>> the definition for test1 is complete
>> the definition test1 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 0;
>>             Tail = 
>>               {<1>}}}}
>> of type: stream(nat)

>> Typing for test2 successful
>> the definition for test2 is complete
>> the definition test2 is provably correct

>> 
>>   {
>>     Head = 1;
>>     Tail = 
>>       {
>>         Head = 1;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 1;
>>                             Tail = 
>>                               {
>>                                 Head = 0;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {<1>}}}}}}}}}}
>> of type: stream(nat)

>> Typing for k3 successful
>> the definition for k3 is complete
>> the definition k3 is provably correct

>> Typing for test3 successful
>> the definition for test3 is complete
>> the definition test3 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 1;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 1;
>>                 Tail = 
>>                   {
>>                     Head = 0;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 1;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 0;
>>                                     Tail = 
>>                                       {<1>}}}}}}}}}}
>> of type: stream(nat)

>> Typing for id successful
>> the definition for id is complete
>> the definition id is provably correct

>> Typing for test4 successful
>> the definition for test4 is complete
>> the definition test4 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 0;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {
>>                                         Head = 1;
>>                                         Tail = 
>>                                           {
>>                                             Head = 0;
>>                                             Tail = 
>>                                               {
>>                                                 Head = 1;
>>                                                 Tail = 
>>                                                   {
>>                                                     Head = 1;
>>                                                     Tail = 
>>                                                       {
>>                                                         Head = 1;
>>                                                         Tail = 
>>                                                           {
>>                                                             Head = 1;
>>                                                             Tail = 
>>                                                               {
>>                                                                 Head = 0;
>>                                                                 Tail = 
>>                                                                   {<1>}}}}}}}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test5 successful
>> the definition for test5 is complete
>> the definition test5 is provably correct

>> 
>>   {
>>     Head = 1;
>>     Tail = 
>>       {
>>         Head = 1;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 1;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {<1>}}}}}}}
>> of type: stream(nat)

>> Typing for test6 successful
>> the definition for test6 is complete
>> the definition test6 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 0;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 0;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 1;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {<1>}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test7 successful
>> the definition for test7 is complete
>> the definition test7 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 0;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {
>>                                         Head = 1;
>>                                         Tail = 
>>                                           {
>>                                             Head = 0;
>>                                             Tail = 
>>                                               {
>>                                                 Head = 1;
>>                                                 Tail = 
>>                                                   {
>>                                                     Head = 1;
>>                                                     Tail = 
>>                                                       {
>>                                                         Head = 1;
>>                                                         Tail = 
>>                                                           {
>>                                                             Head = 1;
>>                                                             Tail = 
>>                                                               {
>>                                                                 Head = 1;
>>                                                                 Tail = 
>>                                                                   {
>>                                                                     Head = 1;
>>                                                                     Tail = 
>>                                                                       {
>>                                                                         Head = 1;
>>                                                                         Tail = 
>>                                                                           {
>>                                                                             Head = 1;
>>                                                                             Tail = 
>>                                                                               {
>>                                                                                 Head = 1;
>>                                                                                 Tail = 
>>                                                                                   {<1>}}}}}}}}}}}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test8 successful
>> the definition for test8 is complete
>> the definition test8 is provably correct

>> 
>>   {
>>     Head = 3;
>>     Tail = 
>>       {
>>         Head = 3;
>>         Tail = 
>>           {
>>             Head = 3;
>>             Tail = 
>>               {
>>                 Head = 3;
>>                 Tail = 
>>                   {
>>                     Head = 3;
>>                     Tail = 
>>                       {
>>                         Head = 3;
>>                         Tail = 
>>                           {
>>                             Head = 3;
>>                             Tail = 
>>                               {
>>                                 Head = 3;
>>                                 Tail = 
>>                                   {
>>                                     Head = 3;
>>                                     Tail = 
>>                                       {
>>                                         Head = 3;
>>                                         Tail = 
>>                                           {<1>}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test9 successful
>> the definition for test9 is complete
>> the definition test9 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 3;
>>         Tail = 
>>           {
>>             Head = 3;
>>             Tail = 
>>               {
>>                 Head = 3;
>>                 Tail = 
>>                   {
>>                     Head = 3;
>>                     Tail = 
>>                       {
>>                         Head = 3;
>>                         Tail = 
>>                           {
>>                             Head = 3;
>>                             Tail = 
>>                               {
>>                                 Head = 3;
>>                                 Tail = 
>>                                   {
>>                                     Head = 3;
>>                                     Tail = 
>>                                       {
>>                                         Head = 3;
>>                                         Tail = 
>>                                           {<1>}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test10 successful
>> the definition for test10 is complete
>> the definition test10 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 2;
>>                 Tail = 
>>                   {
>>                     Head = 3;
>>                     Tail = 
>>                       {
>>                         Head = 4;
>>                         Tail = 
>>                           {
>>                             Head = 5;
>>                             Tail = 
>>                               {
>>                                 Head = 6;
>>                                 Tail = 
>>                                   {
>>                                     Head = 7;
>>                                     Tail = 
>>                                       {
>>                                         Head = 8;
>>                                         Tail = 
>>                                           {<1>}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test12 successful
>> the definition for test12 is complete
>> the definition test12 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 0;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {
>>                                         Head = 1;
>>                                         Tail = 
>>                                           {
>>                                             Head = 0;
>>                                             Tail = 
>>                                               {
>>                                                 Head = 1;
>>                                                 Tail = 
>>                                                   {
>>                                                     Head = 1;
>>                                                     Tail = 
>>                                                       {
>>                                                         Head = 1;
>>                                                         Tail = 
>>                                                           {
>>                                                             Head = 1;
>>                                                             Tail = 
>>                                                               {
>>                                                                 Head = 0;
>>                                                                 Tail = 
>>                                                                   {<1>}}}}}}}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test13 successful
>> the definition for test13 is complete
>> the definition test13 is provably correct

>> 
>>   {
>>     Head = 0;
>>     Tail = 
>>       {
>>         Head = 0;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 0;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 1;
>>                         Tail = 
>>                           {
>>                             Head = 0;
>>                             Tail = 
>>                               {
>>                                 Head = 1;
>>                                 Tail = 
>>                                   {
>>                                     Head = 1;
>>                                     Tail = 
>>                                       {
>>                                         Head = 1;
>>                                         Tail = 
>>                                           {
>>                                             Head = 0;
>>                                             Tail = 
>>                                               {
>>                                                 Head = 1;
>>                                                 Tail = 
>>                                                   {
>>                                                     Head = 1;
>>                                                     Tail = 
>>                                                       {
>>                                                         Head = 1;
>>                                                         Tail = 
>>                                                           {
>>                                                             Head = 1;
>>                                                             Tail = 
>>                                                               {
>>                                                                 Head = 1;
>>                                                                 Tail = 
>>                                                                   {
>>                                                                     Head = 1;
>>                                                                     Tail = 
>>                                                                       {
>>                                                                         Head = 1;
>>                                                                         Tail = 
>>                                                                           {
>>                                                                             Head = 1;
>>                                                                             Tail = 
>>                                                                               {
>>                                                                                 Head = 1;
>>                                                                                 Tail = 
>>                                                                                   {<1>}}}}}}}}}}}}}}}}}}}}}
>> of type: stream(nat)

>> Typing for test14 successful
>> the definition for test14 is complete
>> the definition test14 is provably correct

>> 
>>   {
>>     Head = 1;
>>     Tail = 
>>       {
>>         Head = 1;
>>         Tail = 
>>           {
>>             Head = 1;
>>             Tail = 
>>               {
>>                 Head = 1;
>>                 Tail = 
>>                   {
>>                     Head = 1;
>>                     Tail = 
>>                       {
>>                         Head = 0;
>>                         Tail = 
>>                           {
>>                             Head = 0;
>>                             Tail = 
>>                               {
>>                                 Head = 0;
>>                                 Tail = 
>>                                   {
>>                                     Head = 0;
>>                                     Tail = 
>>                                       {
>>                                         Head = 0;
>>                                         Tail = 
>>                                           {
>>                                             Head = 0;
>>                                             Tail = 
>>                                               {
>>                                                 Head = 0;
>>                                                 Tail = 
>>                                                   {<1>}}}}}}}}}}}}}
>> of type: stream(nat)

